File: game-server/src/main/java/org/forfun/mmorpg/game/scene/director/SceneDirector.java
package org.forfun.mmorpg.game.scene.director;

import org.forfun.mmorpg.game.database.user.entity.PlayerEnt;
import org.forfun.mmorpg.game.scene.actor.ActorType;
import org.forfun.mmorpg.game.scene.actor.Monster;
import org.forfun.mmorpg.game.scene.actor.SceneActor;

public class SceneDirector {

	public void onActorEnter(SceneActor actor) {
		if (actor.getType() == ActorType.Player) {
			onPlayerEnter((PlayerEnt) actor);
		} else if (actor.getType() == ActorType.Monster) {
			onMonsterEnter((Monster) actor);
		}
	}
	
	public void onActorLeave(SceneActor actor) {
		if (actor.getType() == ActorType.Player) {
			onPlayerEnter((PlayerEnt) actor);
		} else if (actor.getType() == ActorType.Monster) {
			onMonsterEnter((Monster) actor);
		}
	}

	protected void onPlayerEnter(PlayerEnt player) {

	}

	protected void onMonsterEnter(Monster monster) {

	}
	
	protected void onPlayerLeave(PlayerEnt player) {

	}

	protected void onMonsterLeave(Monster monster) {

	}

}

File: game-server/src/main/kotlin/org/forfun/mmorpg/game/scene/director/SceneDirector.kt
package org.forfun.mmorpg.game.scene.director

import org.forfun.mmorpg.game.database.user.entity.PlayerEnt
import org.forfun.mmorpg.game.scene.actor.ActorType
import org.forfun.mmorpg.game.scene.actor.Monster
import org.forfun.mmorpg.game.scene.actor.SceneActor

open class SceneDirector {

    fun onActorEnter(actor: SceneActor) {
        if (actor.type == ActorType.Player) {
            onPlayerEnter(actor as PlayerEnt)
        } else if (actor.type == ActorType.Monster) {
            onMonsterEnter(actor as Monster)
        }
    }

    fun onActorLeave(actor: SceneActor) {
        if (actor.type == ActorType.Player) {
            onPlayerLeave(actor as PlayerEnt)
        } else if (actor.type == ActorType.Monster) {
            onMonsterLeave(actor as Monster)
        }
    }

    protected open fun onPlayerEnter(player: PlayerEnt) {}

    protected open fun onMonsterEnter(monster: Monster) {}

    protected open fun onPlayerLeave(player: PlayerEnt) {}

    protected open fun onMonsterLeave(monster: Monster) {}
}
